prov/efa/src/efa_mr.c

```diff
@@ -547,6 +550,139 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
+	ofi_genlock_lock(&efa_mr->domain->util_domain.lock);
+	err = ofi_mr_map_insert(&efa_mr->domain->util_domain.mr_map, mr_attr,
+				&efa_mr->mr_fid.key, &efa_mr->mr_fid);
+	ofi_genlock_unlock(&efa_mr->domain->util_domain.lock);
+
+	/* FI_ENOKEY can happen legitimately when the key was assigned a CUDA memory region, and the memory region has been freed
+	 * after the registration.
+	 *
+	 * This is because of the the combination of following 3 things:
+	 *  1) cuda memory uses memory cache.
+	 *  2) cuda memory cache's monitor is not monitoring call to cudaFree().
+	 *  3) kernel released the device part of memory registration when cudaFree() is called on a registered cuda memory.
+	 *
+	 * Therefore, when application call cudaFree() on a registered CUDA memory region, the device released the device
+	 * part of the memory region, make the key available for reuse. However, libfabric is not aware of the incident,
+	 * and kept the user space memory registration in its MR cache and MR map.
+	 *
+	 * The long term solution to handle this issue is to use a CUDA memory minitor that monitors call to cudaFree()
+	 * (cuda memhook monitor).
+	 * Before that is implemented, we deregister the staled user space memory registration and retry the insert.
+	 *
+	 * TODO: remove this branch once CUDA memhook monitor is implemented and used by efa.
+	 */
+	if (err == -FI_ENOKEY) {
+		existing_mr_fid = ofi_mr_map_get(&efa_mr->domain->util_domain.mr_map, efa_mr->mr_fid.key);
+		assert(existing_mr_fid);
+		existing_mr = container_of(existing_mr_fid, struct efa_mr, mr_fid);
+
+		if (existing_mr->peer.iface != FI_HMEM_CUDA) {
+			EFA_WARN(FI_LOG_DOMAIN, "key %ld already assigned to buffer: %p hmem_iface: %d length: %ld\n",
+				 existing_mr->mr_fid.key,
+				 existing_mr->ibv_mr->addr,
+				 existing_mr->peer.iface,
+				 existing_mr->ibv_mr->length);
+			return -FI_ENOKEY;
+		}
+
+		assert(existing_mr->peer.iface == FI_HMEM_CUDA);
+		err = ofi_cuPointerGetAttribute(&cuda_memory_type, CU_POINTER_ATTRIBUTE_MEMORY_TYPE,
+						(CUdeviceptr)existing_mr->ibv_mr->addr);
+		if (!err) {
+			EFA_WARN(FI_LOG_DOMAIN, "key %ld already assigned to cuda buffer: %p cuda_memor_type: %d length: %ld\n",
+				 existing_mr->mr_fid.key,
+				 existing_mr->ibv_mr->addr,
+				 cuda_memory_type,
+				 existing_mr->ibv_mr->length);
+			return -FI_ENOKEY;
+		}
+
+		EFA_INFO(FI_LOG_DOMAIN, "key %ld has been assigned to cuda buffer: %p length: %ld, which has since been freed\n",
+			 existing_mr->mr_fid.key,
+			 existing_mr->ibv_mr->addr,
+			 existing_mr->ibv_mr->length);
+
+		assert(efa_mr->domain->cache);
+		pthread_rwlock_rdlock(&mm_list_rwlock);
+		pthread_mutex_lock(&mm_lock);
+		ofi_mr_cache_notify(efa_mr->domain->cache, existing_mr->ibv_mr->addr, existing_mr->ibv_mr->length);
+		pthread_mutex_unlock(&mm_lock);
+		pthread_rwlock_unlock(&mm_list_rwlock);
+
+		/* due to MR cache's deferred de-registration, ofi_mr_cache_notify() only move the region to dead_region_list
```

Comment:
I think you should specify why this is important. (IE, the de-register call needs to be triggered since the deregister function has the mr map remove function)

Reply:
adjusted the comment in new revision